src/core/public/chrome/nav_group/nav_group_service.ts

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { BehaviorSubject, combineLatest, Observable, ReplaySubject, Subscription } from 'rxjs';
+import { AppCategory, ChromeNavGroup, ChromeNavLink } from 'opensearch-dashboards/public';
+import { map, takeUntil } from 'rxjs/operators';
+import { IUiSettingsClient } from '../../ui_settings';
+import {
+  flattenLinksOrCategories,
+  fulfillRegistrationLinksToChromeNavLinks,
+  getOrderedLinksOrCategories,
+} from '../utils';
+import { ChromeNavLinks } from '../nav_links';
+
+/** @public */
+export interface ChromeRegistrationNavLink {
+  id: string;
+  title?: string;
+  category?: AppCategory;
+  order?: number;
+
+  /**
+   * link with parentNavLinkId field will be displayed as nested items in navigation.
+   */
+  parentNavLinkId?: string;
+}
+
+export type NavGroupItemInMap = ChromeNavGroup & {
+  navLinks: ChromeRegistrationNavLink[];
+};
+
+export interface ChromeNavGroupServiceSetupContract {
+  addNavLinksToGroup: (navGroup: ChromeNavGroup, navLinks: ChromeRegistrationNavLink[]) => void;
+  /**
+   * Get a boolean value to indicates whether use case is enabled
+   */
+  getNavGroupEnabled: () => boolean;
+}
+
+export interface ChromeNavGroupServiceStartContract {
+  getNavGroupsMap$: () => Observable<Record<string, NavGroupItemInMap>>;
+  getNavGroupEnabled: ChromeNavGroupServiceSetupContract['getNavGroupEnabled'];
+}
+
+/** @internal */
+export class ChromeNavGroupService {
+  private readonly navGroupsMap$ = new BehaviorSubject<Record<string, NavGroupItemInMap>>({});
+  private readonly stop$ = new ReplaySubject(1);
+  private navLinks$: Observable<Array<Readonly<ChromeNavLink>>> = new BehaviorSubject([]);
+  private navGroupEnabled: boolean = false;
+  private navGroupEnabledUiSettingsSubscription: Subscription | undefined;
+  private addNavLinkToGroup(
+    currentGroupsMap: Record<string, NavGroupItemInMap>,
+    navGroup: ChromeNavGroup,
+    navLink: ChromeRegistrationNavLink
+  ) {
+    const matchedGroup = currentGroupsMap[navGroup.id];
+    if (matchedGroup) {
+      const links = matchedGroup.navLinks;
+      const isLinkExistInGroup = !!links.find((link) => link.id === navLink.id);
+      if (isLinkExistInGroup) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          `[ChromeService] Navlink of ${navLink.id} has already been registered in group ${navGroup.id}`
+        );
+        return currentGroupsMap;
+      }
+      matchedGroup.navLinks.push(navLink);
+    } else {
+      currentGroupsMap[navGroup.id] = {
+        ...navGroup,
+        navLinks: [navLink],
+      };
+    }
+
+    return currentGroupsMap;
+  }
+  private getSortedNavGroupsMap() {
```

Comment:
Shall we name it as `getSortedNavGroupsMap$()` since it returns observable?

Reply:
Sure, updated